velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -21,44 +21,135 @@ namespace facebook::velox::connector::hive {
 PartitionIdGenerator::PartitionIdGenerator(
     const RowTypePtr& inputType,
     std::vector<column_index_t> partitionChannels,
-    uint32_t maxPartitions)
+    uint32_t maxPartitions,
+    memory::MemoryPool* pool)
     : partitionChannels_(std::move(partitionChannels)),
       maxPartitions_(maxPartitions) {
-  VELOX_USER_CHECK_EQ(
-      partitionChannels_.size(),
-      1,
-      "Multiple partition keys are not supported yet.");
-  hasher_ = exec::VectorHasher::create(
-      inputType->childAt(partitionChannels_[0]), partitionChannels_[0]);
+  VELOX_USER_CHECK(
+      !partitionChannels_.empty(), "There must be at least one partition key.");
+  for (auto channel : partitionChannels_) {
+    hashers_.emplace_back(
+        exec::VectorHasher::create(inputType->childAt(channel), channel));
+  }
+
+  std::vector<TypePtr> partitionKeyTypes;
+  for (auto channel : partitionChannels_) {
+    partitionKeyTypes.push_back(inputType->childAt(channel));
+  }
+
+  partitionKeys_ = BaseVector::create<RowVector>(
+      ROW(std::move(partitionKeyTypes)), maxPartitions_, pool);
+  for (auto& key : partitionKeys_->children()) {
+    key->resize(maxPartitions_);
+  }
 }
 
 void PartitionIdGenerator::run(
     const RowVectorPtr& input,
     raw_vector<uint64_t>& result) {
-  result.resize(input->size());
+  auto numRows = input->size();
+  result.resize(numRows);
+
+  // TODO Check that there are no nulls in the partition keys.
+
+  // Compute value IDs using VectorHashers and store these in 'result'.
+  computeValueIds(input, result);
+
+  // Convert value IDs in 'result' into partition IDs using partitionIds
+  // mapping. Update 'result' in place.
+
+  // TODO Optimize common use case where all records belong to the same
+  // partition. VectorHashers keep track of the number of unique values, hence,
+  // we can find out if there is only one unique value for each partition key.
+  for (auto i = 0; i < numRows; ++i) {
+    auto valueId = result[i];
+    auto it = partitionIds_.find(valueId);
+    if (it != partitionIds_.end()) {
+      result[i] = it->second;
+    } else {
+      uint64_t nextPartitionId = partitionIds_.size();
+      VELOX_USER_CHECK_LT(
+          nextPartitionId,
+          maxPartitions_,
+          "Exceeded limit of distinct partitions.");
+
+      partitionIds_.emplace(valueId, nextPartitionId);
+      savePartitionKeys(nextPartitionId, input, i);
+
+      result[i] = nextPartitionId;
+    }
+  }
+
+  recentMaxId_ = *std::max_element(result.begin(), result.end());
```

Comment:
What's use of recentMaxId_? Thanks!

Reply:
@xiaoxmeng @gggrace14 I'm not sure what recentMaxId_ is used for. It may not be needed. Ge, do you know?